rpc/handlers.go

```diff
@@ -878,21 +887,35 @@ func (h *Handler) AddInvokeTransaction(invokeTx json.RawMessage) (*AddInvokeTxRe
 	return invokeRes, nil
 }
 
-// getAddInvokeTxCode returns the relevant Code for a given addInvokeTx error
-func getAddInvokeTxCode(err error) int {
-	switch {
-	case strings.Contains(err.Error(), "contract address") && strings.Contains(err.Error(), "is out of range"):
-		return jsonrpc.InvalidParams
-	case strings.Contains(err.Error(), "Fee") && strings.Contains(err.Error(), "is out of range"):
-		return jsonrpc.InvalidParams
-	case strings.Contains(err.Error(), "Missing data for required field"):
-		return jsonrpc.InvalidParams
-	case strings.Contains(err.Error(), "not supported. Supported versions"):
-		return jsonrpc.InvalidParams
-	case strings.Contains(err.Error(), "max_fee must be bigger than 0"):
-		return jsonrpc.InvalidParams
+func makeJSONErrorFromGatewayError(err error) *jsonrpc.Error {
+	gatewayErr, ok := err.(*gateway.Error)
+	if !ok {
+		return jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+
+	switch gatewayErr.Code {
+	case gateway.InvalidContractClass:
+		return ErrInvalidContractClass
+	case gateway.UndeclaredClass:
+		return ErrClassHashNotFound
+	case gateway.ClassAlreadyDeclared:
+		return ErrClassAlreadyDeclared
+	case gateway.InsufficientMaxFee:
+		return ErrInsufficientMaxFee
+	case gateway.InsufficientAccountBalance:
+		return ErrInsufficientAccountBalance
+	case gateway.ValidateFailure:
+		return ErrValidationFailure
+	case gateway.ContractBytecodeSizeTooLarge:
+		return ErrContractBytecodeSizeTooLarge
+	case gateway.DuplicatedTransaction:
+		return ErrDuplicateTx
+	case gateway.InvalidTransactionNonce:
+		return ErrInvalidTransactionNonce
+	case gateway.CompilationFailed:
+		return ErrCompilationFailed
 	default:
-		return jsonrpc.InternalError
+		return jsonrpc.Err(jsonrpc.InvalidParams, err.Error())
```

Comment:
So since `err.(*gateway.Error)` succeeded, gateway returned us a proper error message. Which means that Juno fulfilled it's job a proxy. We can assume it is an error related to the request rather than an internal error that is related to Juno.